### PR TITLE
Scoped secrets

### PR DESCRIPTION
**Pull request by stefannica**
## Add secret scope support to Secrets Managers

The majority of Secrets Managers use a single global scope in the back-end they are connected to, meaning that all secrets configured and visible through a Secrets Manager instance are also visible in all other instances. Furthermore, secrets configured directly in the back-end through means other than ZenML are also visible and can be managed in all ZenML Secrets Managers connected to that same back-end. This has a number of disadvantages:

* there is only one global shared namespace for secret names. Two secrets with the same name and different values cannot be configured at the same time for different purposes (e.g. different ZenML projects or components).
* there is no clear separation between secrets managed by ZenML and those that are used for other purposes

This PR introduces hierarchical scoping to Secrets Managers.

### Secret Manager API and Behavior Changes

All Secrets Managers get a set of new configuration attributes that determine how and if a Secrets Manager instance shares secrets with other Secrets Manager instances connected to the same back-end:

* `scope` determines the secret scope and can be set to one of the following values:
    * `none`: no secret scoping is used when this scope value is configured (i.e. the existing behavior prior to this PR). It is only used to preserve backwards compatibility with existing Secrets Manager instances that do not yet understand the concept of secrets scoping. Will be deprecated and phased out in future ZenML versions. 
    * `global`: secrets are shared across all Secrets Manager instances that connect to the same backend and have a global scope.
    * `component`: secrets are not shared outside a Secrets Manager instance. This is the default for new Secrets Manager instances.
    * `namespace`: secrets are shared only by Secrets Manager instances that connect to the same backend *and* have the same `namespace` scope value configured (see below).

* `namespace` is a scope namespace value to use with the `namespace` scope

Every Secrets Manager maps the scope differently to how secrets are named and stored in the back-end, but if the scope could be converted into a path or name prefix, a secret named `aria` would be stored like this:

* in the `none` scope, it would be stored as `aria` (no name prefix) in the root path
* in the `global` scope, it would be stored as `zenml/global/aria`
* in the `component` scope, it would be stored as `zenml/component/<component-UUID>/aria`
* in the `namespace` scope, it would be stored ad `zenml/namespace/<namespace-value>/aria`

### AWS Secret Manager Changes

The AWS Secrets Manager is updated to support secret scoping by converting the scope into an AWS secret name prefix.

Other AWS Secrets Manager changes:

* secret name validation (only some characters are permitted)
* adds tags to AWS secrets to help with visibility and to be used as query filtering criteria
* better scalability by not fetching all the available secrets to check if a single secret exists
* integration tests

### GCP Secret Manager Changes

The GCP Secrets Manager is updated to support secret scoping by converting the scope into an GCP secret name prefix.

Other GCP Secrets Manager changes:

* secret name validation (only some characters are permitted)
* adds labels to GCP secrets to help with visibility and to be used as query filtering criteria
* better scalability by not fetching all the available secrets to check if a single secret exists
* integration tests

### Vault Secret Manager Changes

The Vault Secrets Manager is updated to support secret scoping by converting the scope into a secret name prefix.

Other Vault Secrets Manager changes:

* secret name validation (only some characters are permitted)
* integration tests

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/advanced-guide/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/integrations/aws/secrets_managers/aws_secrets_manager.py [hunk 1]
```diff
@@ -71,6 +90,75 @@ def _ensure_client_connected(cls, region_name: str) -> None:
...
+            name: the name of the AWS scoped secret
+
+        Returns:
+            The ZenML secret name or None, if the input secret name does not
+            belong to the current scope.
+        """
+        name_tokens = name.split("/")
+        secret_name = name_tokens[-1]
+        name_tokens[-1]
+        prefix = "/".join(name_tokens)
```
Comment (schustmi) [inline src/zenml/integrations/aws/secrets_managers/aws_secrets_manager.py:154]:
```suggestion
        prefix = "/".join(name_tokens[:-1])
```

**Comment by RoyerRamirez** [comment]
Hi @stefannica, @schustmi , @AlexejPenner  I hope all is well. This is a great idea, and we faced similar issues as well. Would it be possible to use the following naming convention instead?

* `none` scope:
`zenml/aria`

* `global` scope:
`zenml/global/aria`

* `component` scope:
`zenml/component/<component-UUID>/aria`

* `namespace` scope:
`zenml/namespace/<namespace>/aria`

Reasons for starting all secrets with `zenml`:
* When managing a large cloud environment it's important to know which secrets belong to which cloud component. By starting every secret that is "owned" by ZenML, with `zenml/`, then it's very clear what application owns that secret. 
* AWS and GCP also gives us the flexibility to breakdown permissions based on aliases. For example, I may want to grant an application access to all zenml secrets for auditing purposes. If I don't start every secret with `zenml/` I won't know all the secrets that are owned by ZenML unless I go into each application's code. 

Reasons for being more explicit with the secret name:
* Placing the word `component` and `namespace` in the actual name will help identify the secret correctly. 
* Users don't have to guess if a ZenML secret is referring to a namespace or component
* Google Cloud does this when they define an item. For example, this is how Google defines a vm resources `projects/{{project}}/zones/{{zone}}/instances/{{name}}`. ZenML should do something similar since it makes it very clear what each item is. 

Best,

Royer R Ruiz

File: src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py [hunk 2]
```diff
@@ -105,14 +297,15 @@ def register_secret(self, secret: BaseSecretSchema) -> None:
         Raises:
             SecretExistsError: if the secret already exists
         """
+        self.validate_secret_name_or_namespace(secret.name)
         self._ensure_client_connected()
 
-        if secret.name in self.get_all_secret_keys():
+        if not self._list_secrets(secret.name):
```
Comment (schustmi) [inline src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py:303]: Shouldn't this be without the `not` in order to work?

**Comment by stefannica** [comment]
> Hi @stefannica, @schustmi , @AlexejPenner I hope all is well. This is a great idea, and we faced similar issues as well. 

Hi @RoyerRamirez , this is great feedback and easily achievable with a few minor changes to the existing implementation. The only part I disagree with for technical reasons is the `none` scope, which is, as the PR description says, only used for backwards compatibility and needs to preserve the previous naming schema.

File: src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py [hunk 3]
```diff
@@ -176,102 +371,135 @@ def get_secret(self, secret_name: str) -> BaseSecretSchema:
...
+                google_secret = self.CLIENT.get_secret(name=google_secret_name)
+            except google_exceptions.NotFound:
+                raise KeyError(
+                    f"Can't find the specified secret '{secret_name}'"
+                )
+
+            # make sure the secret has the correct scope labels to filter out
+            # unscoped secrets with similar names
+            scope_labels = self._get_secret_scope_metadata(secret_name)
+            if not scope_labels.items() <= google_secret.labels.items():
```
Comment (schustmi) [inline src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py:411]: I have never seen that syntax of comparing dict items, can you explain what it does here?

File: src/zenml/integrations/vault/secrets_manager/vault_secrets_manager.py [hunk 4]
```diff
@@ -149,11 +79,44 @@ def _ensure_client_is_authenticated(self) -> None:
...
+    def _sanitize_secret_name(secret_name: str) -> str:
+        """Sanitize the secret name to be used in Vault.
+
+        Args:
+            secret_name: The secret name to sanitize.
+
+        Returns:
+            The sanitized secret name.
+        """
+        sanitized_secret_name = re.sub(r"[^0-9a-zA-Z_\.]+", "_", secret_name)
```
Comment (schustmi) [inline src/zenml/integrations/vault/secrets_manager/vault_secrets_manager.py]: In the other secrets manager we simply prevent unallowed special characters, should we do the same thing here?

File: src/zenml/secrets_managers/base_secrets_manager.py [hunk 5]
```diff
@@ -14,24 +14,299 @@
...
+    scope: SecretsManagerScope = SecretsManagerScope.COMPONENT
+    namespace: Optional[str] = None
+
+    @root_validator(pre=True)
+    def scope_initializer(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """Pydantic root_validator for the scope.
+
+        This root validator is used for backwards compatibility purposes. It
+        ensures that existing Secrets Managers continue to use no scope for
+        secrets while new instances default to using the component scope.
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/secrets_managers/base_secrets_manager.py:78]: I thought about it and I would keep the UUIDs out of the stack import/export, for the simple reason that there are specific scope configurations that you can use to share secrets (global or namespace). Sharing stack components via a shared profile configuration or the ZenML Server is another way of addressing this. I view the stack import/export more as a way of sharing stack configuration templates than the actual stack configurations.

**Review by schustmi — CHANGES_REQUESTED** [review]

File: src/zenml/secrets_managers/base_secrets_manager.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply src/zenml/secrets_managers/base_secrets_manager.py:78]: Maybe we should add a line about this in the docs then, 

`Using the component scope means that only stacks with a secret manager with the exact UUID as your local stack can access these secrets. The global or namespace scope are more suitable for shared access to secrets`

File: src/zenml/secrets_managers/base_secrets_manager.py [hunk 6]
```diff
@@ -14,24 +14,299 @@
...
+        Returns:
+            Values passed to the object constructor
+
+        Raises:
+            ValueError: If the scope value is not valid.
+        """
+        scope = values.get("scope")
+
+        # this is a new Secrets Manager instance
+        if "uuid" not in values:
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/secrets_managers/base_secrets_manager.py:92]: This is the only way to differentiate between secrets manager instances created prior to zenml 0.12.0, that use the old unscoped naming schema and the new instances that are created post 0.12.0. But I do need to rethink the updating part, you're correct that those checks also apply.

File: src/zenml/integrations/vault/secrets_manager/vault_secrets_manager.py (hunk 4 above)
Comment (AlexejPenner) [reply src/zenml/integrations/vault/secrets_manager/vault_secrets_manager.py]: Fully agree, should have been replaced already with https://github.com/zenml-io/zenml/pull/803/commits/82f444a3a2e58c1b92b78f2766d91de26138a34d - not sure why it's not showing up in the files diff yet

File: src/zenml/utils/string_utils.py [hunk 7]
```diff
@@ -100,4 +100,5 @@ def random_str(length: int) -> str:
     Returns:
         Random human-readable string.
     """
+    random.seed()
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/utils/string_utils.py:103]: Yes, that is exactly the conclusion I came to: something in the AWS SDK code uses a fixed seed somewhere.

Normally, I would agree that we need to respect the fixed seed, but in this case that comes into conflict with ZenML code. More than that, I see using a fixed seed in production code as a potential security risk and I even spent a considerable amount of time debugging it and trying to understand where exactly in the AWS SDK a fixed seed is used, but could not find it. This fix will have to do for the time being.

File: src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py [hunk 8]
```diff
@@ -96,6 +142,152 @@ def parent_name(self) -> str:
...
+    def _get_scoped_secret_name(self, name: str) -> str:
+        """Convert a ZenML secret name into a Google scoped secret name.
+
+        Args:
+            name: the name of the secret
+
+        Returns:
+            The Google scoped secret name
+        """
+        return "-".join(self._get_scoped_secret_path(name))
```
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py:154]: I thought of that too, but this is just a few extra characters vs. creating a new function with docstrings and everything around it, which will amount to many lines of code.

[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py:154]: On third thought, this seems more sensible now that we have 3 secrets managers that use this type of logic.

[the remark this answers is not in the document]
Comment (schustmi) [reply src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py:154]: Yep not something that we need to do right now, potentially once we tackle the remaining secrets managers

File: src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py (hunk 2 above)
Comment (stefannica) [reply src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py:303]: Curiously enough, I just ran the integration tests and they also caught this same issue, but they passed just fine the last time I ran them. Probably something that slipped during a merge. Thanks for noticing !

File: src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py (hunk 3 above)
Comment (stefannica) [reply src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py:411]: This is some python 3 magic. It checks that all the items in the scope_labels dictionary are included in the google secrets labels. I added a comment to clarify.

The `dict_items` type behaves like a set: all of the operations defined for the abstract base class collections.abc.Set are available (for example, ==, <, or ^). More here: https://docs.python.org/3/library/stdtypes.html#dict-views

File: src/zenml/integrations/vault/secrets_manager/vault_secrets_manager.py (hunk 4 above)
Comment (stefannica) [reply src/zenml/integrations/vault/secrets_manager/vault_secrets_manager.py]: @AlexejPenner will take care of this :)

File: src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py (hunk 3 above)
Comment (schustmi) [reply src/zenml/integrations/gcp/secrets_manager/gcp_secrets_manager.py:411]: I see, a set of tuples and then it's a subset check, interesting. Thanks!

File: docs/book/mlops-stacks/secrets-managers/secrets-managers.md [hunk 9]
```diff
@@ -156,3 +156,73 @@ their default value if omitted
 * all values must be a valid string representation of the data type indicated in
 the schema (i.e. that can be converted to the type indicated) or an error will
 be raised
+
+## Secret Scopes
```
[the remark this answers is not in the document]
Comment (stefannica) [reply docs/book/mlops-stacks/secrets-managers/secrets-managers.md:160]: Yes, a visualization would definitely help. I'll see if I can prepare something.

File: src/zenml/secrets_managers/base_secrets_manager.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefannica) [reply src/zenml/secrets_managers/base_secrets_manager.py]: Done

**Comment by stefannica** [comment]
@schustmi @AlexejPenner aside from the visualization, I think I addressed all your code review comments.

**Review by schustmi — APPROVED** [review]
I really like how you implemented this PR, was a pleasure to read/review! 🙂

**Review by AlexejPenner — APPROVED** [review]
LTSM - Let the Stefan Merge

**Pull request merged**